### ISSUE TITLE
input validation: reject only-zero samples?

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#532]
Here, a single zero was submitted:

![image](https://user-images.githubusercontent.com/265630/207322586-8330d282-6497-4446-b995-b2762eea3e67.png)

Maybe this should not survive sanity data validation check, and be rejected with a helpful error message wrapped in a 400 Bad Request response.

**Comment by jgehrcke** [comment]
https://conbench.ursa.dev/benchmarks/df3b604775524876aa1ac9a03cf2b3ad/

![image](https://user-images.githubusercontent.com/265630/223660436-f5ee3ef0-e6fb-4735-b522-d8b1dde45106.png)

It's likely that this is a methodological error. If Conbench were to reject this it's easier to discover.

If this is not a methodological error then I'd be curious to learn what the value of this result is.

**Comment by jonkeane** [comment]
Agreed, separately we should reach out to the Arrow Javascript folks and ask them about this.

We might consider doing this error|warning in the client (benchadapt or benchclient) first to minimize disruptiveness: start off with a warning or error there but still accept the (bad) info into conbench. We would be able to provide slightly more context at the client level + see if that's enough nudge before baking the restriction in to the server. 

I will also say, IMO, looking at that data in conbench it's much easier to see how not-helpful it is compared to the the command line | getting a 400 (even with a specific message)

**Comment by jgehrcke** [comment]
A few more thoughts:

- This is very likely a mean bug in some client/benchmarking tooling. Might have to do with string truncation, even. `0` means infinitely fast or infinitely slow -- I just thought about this again, this can never be a good data point I think. If I were to write tooling that submits these values to Conbench -- I'd _love_ for Conbench to reject my request, to tell me that my tool chain has a bug.
- The advantage of rejecting this would be to let the CI job fail. A failed CI job is more likely to be debugged in a timely fashion than a succeeding one. Enforced communication :-).
- The API server can generate a precise, nice, simple error message.
- Our client tooling should expose a that error message to the user. If it does not, we should also work on that.

But I also see the point that if manyyyyy (independently obtained) results are submitted in the same request, that it might hurt the process a lot if the entire submission request fails. So in that case we may want to partially accept the good data, and reject only the individual bad result -- and _still_ send an error response.

**Comment by jgehrcke** [comment]
I think I might have found the relevant code, somewhere around here:
https://github.com/apache/arrow/blob/71f3c568af8fe8a6f886ffddb4318b728046a01a/js/perf/index.ts#L60
at least there's quite a bit of string formatting going on.

CC @domoritz

**Comment by jonkeane** [comment]
> But I also see the point that if manyyyyy (independently obtained) results are submitted in the same request, that it might hurt the process a lot if the entire submission request fails. So in that case we may want to partially accept the good data, and reject only the individual bad result -- and still send an error response.

Yeah, this is very important. IIRC, these benchmarks are in that camp: all the results are written out and then we upload them all at once. It would be unfortunate if we rejected 73 benchmarks because 15 of them had an issue (If I'm looking at the javascript benchmarks correctly, that's the ratio we're talking about, but it could be hundreds of good ones and a single one that has 0 erroneously).

I share your suspicion that this is likely something funky with either the way we're running | munging the data coming out of the benchmark framework that the javascript benchmarks use or something in the javascript benchmark framework itself. We should _definitely_ reach out to the javascript maintainers and see if they have any ideas about these particular cases — that will at least help figure out where the disconnect is happening.

**Comment by jgehrcke** [comment]
> all the results are written out and then we upload them all at once

and they often are largely independent. Hence, a single HTTP request might contain a collection of independently obtained benchmark results -- this indeed requires special consideration. I opened https://github.com/conbench/conbench/issues/856 with an initial thought dump.

**Comment by domoritz** [comment]
The JSON results (which I thought was the only thing being submitted) is generated in https://github.com/apache/arrow/blob/71f3c568af8fe8a6f886ffddb4318b728046a01a/js/perf/index.ts#L55 before the formatting. Or do you see 0 runtime when you run the benchmark manually?
